File: example-crs-webservice/crs-userspace/libs/libMSA/python/tests/thread/test_equeue.py
import random
import threading
import time
import unittest
from typing import List

from libmsa import EventfulQueue


class TestEventfulQueue(unittest.TestCase):
    def setUp(self) -> None:
        self.work_queue = EventfulQueue()
        self.terminate = False

    def enqueue_worker(self):
        for _ in range(100000):
            data = random.randint(0, 100000)
            self.work_queue.enqueue(data)

    def dequeue_worker(self):
        event = self.work_queue.get_event()
        while not self.terminate or event.is_set():
            event.wait()

            while True:
                data = self.work_queue.dequeue()
                if data is None:
                    break
                time.sleep(0.0001)
            self.work_queue.task_done()

    def test_single_thread_basic(self):
        data = 1

        self.assertFalse(self.work_queue.get_event().is_set())
        self.work_queue.enqueue(data)
        self.assertTrue(self.work_queue.get_event().is_set())
        self.work_queue.task_done()
        self.assertTrue(self.work_queue.get_event().is_set())
        ret = self.work_queue.dequeue()
        self.assertEqual(data, ret)
        ret = self.work_queue.dequeue()
        self.assertIsNone(ret)
        self.work_queue.task_done()
        self.assertFalse(self.work_queue.get_event().is_set())

    def test_multiple_thread_basic(self):
        num_enqueue_threads = 1
        num_dequque_threads = 10

        enqueue_threads: List[threading.Thread] = []
        dequeue_threads: List[threading.Thread] = []

        for _i in range(num_enqueue_threads):
            thread = threading.Thread(target=self.enqueue_worker, args=())
            enqueue_threads.append(thread)
            thread.start()

        for _ in range(num_dequque_threads):
            thread = threading.Thread(target=self.dequeue_worker, args=())
            dequeue_threads.append(thread)
            thread.start()

        for thread in enqueue_threads:
            thread.join()

        self.terminate = True

        for thread in dequeue_threads:
            thread.join()

        self.assertIsNone(self.work_queue.dequeue())


if __name__ == "__main__":
    unittest.main()

File: example-crs-webservice/crs-userspace/libs/libMSA/cpp/tests/thread/test_equeue.cpp
#include <gtest/gtest.h>
#include <thread>
#include <chrono>
#include "libmsa/thread/equeue.h"
#include "test.pb.h"

class TestEventfulQueue : public ::testing::Test {
protected:
    EventfulQueue<userspace::TestRequest> work_queue;
    std::atomic<bool> terminate;
    const char* payload;

    void SetUp() override {
        terminate = false;
        payload = "Team-Atlanta";
    }

    void TearDown() override {

    }
public:
    void enqueue_worker() {
        for (int i = 0; i < 100000; ++i) {
            userspace::TestRequest request;
            request.set_team_name(payload);
            work_queue.enqueue(&request);
        }
    }

    void dequeue_worker() {
        while (!terminate.load()) {
            for(;;){
                userspace::TestRequest* data = work_queue.dequeue();
                if (data == nullptr) {
                    break;
                }
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(1));
        }
    }
};

TEST_F(TestEventfulQueue, SingleThreadBasic) {
    userspace::TestRequest* request;
    userspace::TestRequest* dequeued_data;

    dequeued_data = work_queue.dequeue();
    EXPECT_EQ(dequeued_data, nullptr);

    request = new userspace::TestRequest();
    request->set_team_name(payload);

    work_queue.enqueue(request);

    delete request;

    dequeued_data = work_queue.dequeue();
    EXPECT_EQ(dequeued_data->team_name(), payload);
    work_queue.task_done(dequeued_data);

    dequeued_data = work_queue.dequeue();
    EXPECT_EQ(dequeued_data, nullptr);
}

TEST_F(TestEventfulQueue, MultipleThreadBasic) {
    const int num_enqueue_threads = 1;
    const int num_dequeue_threads = 10;

    std::vector<std::thread> enqueue_threads;
    std::vector<std::thread> dequeue_threads;

    for (int i = 0; i < num_enqueue_threads; ++i) {
        enqueue_threads.emplace_back(&TestEventfulQueue::enqueue_worker, this);
    }

    for (auto& thread : enqueue_threads) {
        thread.join();
    }

    for (int i = 0; i < num_dequeue_threads; ++i) {
        dequeue_threads.emplace_back(&TestEventfulQueue::dequeue_worker, this);
    }

    terminate.store(true);

    for (auto& thread : dequeue_threads) {
        thread.join();
    }

    EXPECT_EQ(work_queue.dequeue(), nullptr);
}

int main(int argc, char** argv) {
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}